pandas/tests/reshape/concat/test_index.py

```diff
@@ -265,3 +265,40 @@ def test_concat_multiindex_dfs_with_deepcopy(self):
         tm.assert_frame_equal(result_copy, expected)
         result_no_copy = concat(example_dict, names=["testname"])
         tm.assert_frame_equal(result_no_copy, expected)
+
+    @pytest.mark.parametrize(
+        "mi1",
+        [
+            MultiIndex.from_product([["a"], range(2)]),
+            MultiIndex.from_product([["a"], np.arange(2.0)]),
+            MultiIndex.from_product([["b"], ["A", "B"]]),
+            MultiIndex.from_product(
+                [["c"], pd.date_range(start="2017", end="2018", periods=2)]
+            ),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "mi2",
+        [
+            MultiIndex.from_product([["a"], range(2)]),
+            MultiIndex.from_product([["a"], np.arange(2.0)]),
+            MultiIndex.from_product([["b"], ["A", "B"]]),
+            MultiIndex.from_product(
+                [["c"], pd.date_range(start="2017", end="2018", periods=2)]
+            ),
+        ],
+    )
+    def test_concat_with_various_multiindex_dtypes(self, mi1, mi2):
+        # GitHub #23478
+        df1 = DataFrame(np.zeros((1, len(mi1))), columns=mi1)
+        df2 = DataFrame(np.zeros((1, len(mi2))), columns=mi2)
+
+        expected = DataFrame(
+            np.zeros((1, len(mi1) + len(mi2))),
+            columns=MultiIndex.from_tuples(list(mi1) + list(mi2)),
```

Comment:
Regarding you comment, instead could you instruct this with the `levels` and `codes` in the regular constructor instead?